### ISSUE TITLE
Upgrade hyperHTML to 2.0

### ISSUE THREAD

**Issue by robdodson** [webcomponents/custom-elements-everywhere#139]
@WebReflection I tried updating to `2.0.6` but noticed all the tests failed because

```
import {hyper} from 'hyperhtml'
```

no longer works. The docs mention that `hyper` is still an exported function so I must be doing something wrong with my use of modules. Any ideas?

**Comment by WebReflection** [comment]
I am not sure what could be the issue but hyperHTML is exported as real ESM and you can test it directly through [this page](https://webreflection.github.io/hyperHTML/test/esm.html) which content is:

```html
<!doctype html>
<script type=module>
import {hyper} from '../esm/main.js';
window.hyper = hyper;
</script>
```

**However**, I was using `global` references that my bundlers where happily ignoring/converting but not on the ESM source code so, in case your tools used the `module` property in package.json without expecting a `global`, that might have been an issue.

The regular CommonJS version also export `hyperHTML` main function as default, but also as module `hyper` property and recursively set `hyper` property to the `hyper` function so there's no way you shouldn't be able to import that.

As last resort, try `import {hyper} from 'hyperhtml'` and see how that goes.

**Comment by WebReflection** [comment]
Hey @robdodson not sure you had time to check anything but here the thing ... beside the performance release I've done today (and I promise I'll be quite with releases for a little while now) I've PR'd another repository that wasy using rollup which is apparently incapable of resolving hyperhtml, even if the module in the package.json is correct.

TL;DR can you try, with latest hyperHTML (v.2.1.1) to do the following?

```js
import hyper from 'hyperhtml/esm/main.js'
```

if that works, I've already configured my repo to use `index.js` files instead of main so that in the future all you need to do would be:

```js
import hyper from 'hyperhtml/esm'
```

and eventually

```js
const {hyper} = require('hyperhtml/cjs');
```

so that everything should work as expected, no matter where you import/require it.

Thanks for your patience in keeping hyperHTML updated

**P.S.** version 2 introduced changes to the template logic so that Custom Elements are always upgraded and the async hook should not be needed anymore. I'll have a look into testing this once you've managed to update the dependency

**Comment by WebReflection** [comment]
@robdodson 2.1.2 provides esm/index so you can `import {hyper} from 'hyperhtml/esm'`

**Comment by robdodson** [comment]
That seems to work. Though it looks like I still have some broken tests complaining about a missing hyperHTML.bind. That may me a mistake on my end though. I'm heading out for vacation in a few hours so I have to run, but I'll see if I can check back in on this in a few days.

For now the current site just shows v1.12.5 which gets a 30/30 :)

**Comment by WebReflection** [comment]
`import hyper from 'hyperhtml/esm'` should bring the default export which has `bind` in it, but you could `import {hyper, bind} from 'hyperhtml/esm'` too.

let me know how that goes when you'll have time

**Comment by robdodson** [comment]
It seems to be failing on [this line in hyperHTML-Element](https://github.com/WebReflection/hyperHTML-Element/blob/master/index.js#L167).

Stacktrace:
```
	hyperHTML.bind is not a function
	get html@webpack:///node_modules/hyperhtml-element/index.js:167:0 <- tests.webpack.js:6992:42
	render@webpack:///src/components.js:120:4 <- tests.webpack.js:5686:5
	created@webpack:///src/components.js:112:14 <- tests.webpack.js:5667:5
	value@webpack:///node_modules/hyperhtml-element/index.js:87:0 <- tests.webpack.js:6912:17
	Cd.prototype.connectedCallback@node_modules/@webcomponents/webcomponentsjs/webcomponents-lite.js:134:194
	M@node_modules/@webcomponents/webcomponentsjs/webcomponents-lite.js:129:908
	ue/<@node_modules/@webcomponents/webcomponentsjs/webcomponents-lite.js:145:382
	@webpack:///src/component-tests.js:206:4 <- tests.webpack.js:2092:5
```

**Comment by WebReflection** [comment]
sorry @robdodson it looks like I had to work a bit more on HyperHTMLElement to make it compatible with v2 ... it should be the case now, if you have issues with v1 try to import it via:

```js
import HyperHTMLElement from 'hyperhtml-element/esm';
```

you might need a `import alias from "rollup-plugin-alias";` in rollup to then add the dependency:
```js
alias({
  'https://unpkg.com/hyperhtml@latest/esm/index.js': 'node_modules/hyperhtml/esm/index.js'
})
```

**or** you can try via
```js
import HyperHTMLElement from 'hyperhtml-element/cjs';
```

which is the CommonJS with the default solution ... too bad tools aren't really smarts these days with real ESM modules :disappointed:

**Comment by WebReflection** [comment]
@robdodson I've just published [rollup-plugin-cdn](https://github.com/WebReflection/rollup-plugin-cdn) which should solve automagically all your issues with modules if you are using rollup.

Just add the `cdn()` plugin and no manual alias should be ever needed again :tada: 
```js
// rollup.config.js
import cdn from 'rollup-plugin-cdn';

export default {
  input: './esm/index.js',
  plugins: [cdn()],
};
```

**Comment by WebReflection** [comment]
... Rob ?

**Comment by robdodson** [comment]
Hey sorry, been traveling for a bit :)

I'm using Webpack and it seems like it's getting hung up on the line in HyperHTMLElement that tries to import from unpkg.

```
import HyperHTMLELement from 'hyperhtml-element/esm';
```

Gives the following error:

```
Module not found: Error: Can't resolve 'https://unpkg.com/hyperhtml@latest/esm/index.js' in '/Users/.../custom-elements-everywhere/libraries/hyperhtml/node_modules/hyperhtml-element/esm'
```

Maybe Webpack is not able to pull in modules from external URLs?

**Comment by WebReflection** [comment]
can you please do this instead then?

```js
import HyperHTMLElement from 'hyperhtml-element/cjs';
```

**Comment by robdodson** [comment]
Just tried that but it seems to result in the missing `bind` error again :\

```
Error: Uncaught TypeError: bind is not a function (webpack:///node_modules/hyperhtml-element/cjs/index.js:176:0
```

**Comment by WebReflection** [comment]
what the hell is wrong with Webpack ? and what happens if you do this ?

```js
import HyperHTMLElement from 'hyperhtml-element';
```

also, please consider updating `HyperHTMLElement` to the latest and do not include `hyperhtml` at all since latest `HyperHTMLElement` ships with `hyperhtml` already.

I might re-fork and PR latest myself though so you won't waste your time.

**Comment by robdodson** [comment]
currently i'm doing:

```
import {hyper} from 'hyperhtml/cjs';
import HyperHTMLELement from 'hyperhtml-element/cjs';
```

which results in:

```
bind is not a function
	get html@webpack:///node_modules/hyperhtml-element/cjs/index.js:176:0
```

> also, please consider updating HyperHTMLElement to the latest
locally I'm on 1.1.0

> do not include hyperhtml at all since latest HyperHTMLElement ships with hyperhtml already

I was thinking of trying this as well. I'll need to redo a couple of the elements so I was being lazy :)

**Comment by WebReflection** [comment]
I had to change 3 times the export to make Webpack happy in your configuration.

This wasn't good UX with Webpack, IMO, but I'm happy I've eventually solved the issue:
https://github.com/webcomponents/custom-elements-everywhere/pull/151

The problem is that my module is now compromised and it cannot be used via CDN because of Webpack issues.

I will address these issues with Webpack people.

I think you can close this PR and consider mine instead. Thanks for taking the time to test this.
